### PR TITLE
Closes #186 Support for bulk question import from csv 

### LINKED ISSUE #186: Add Support for Bulk Question Import via CSV in Automated Test Suite
#### Summary:
Our AI project, Ayushma, currently relies on a manual process for entering questions into our automated test suite. This approach is increasingly cumbersome and inefficient, especially as we scale up our testing with larger datasets. To streamline this process, we propose the implementation of a feature that allows for the bulk import of questions via a CSV file.

#### Feature Request:
**Automated Question Import via CSV**

#### Description:
The automated test suite should be enhanced to support the import of questions in bulk from a CSV file. This feature would significantly reduce the time and effort required to populate the test suite with a large number of questions, facilitating more efficient testing cycles and allowing for a broader range of tests to be conducted in a shorter time frame.

#### Proposed CSV Format:
The CSV file should adhere to the following format for each entry:
```
question, human_answer, language_code
```
- **question**: The question text to be tested.
- **human_answer**: The expected answer or response that a human would provide.
- **language_code**: The ISO language code representing the language of the question and answer (e.g., `en` for English).

#### Implementation Details:
- The test suite should include a function or interface for importing questions from a CSV file.
- Upon import, the suite should parse the CSV file and add each question to the testing queue, along with its corresponding human-provided answer and language code.
- The feature should support error handling for improperly formatted CSV files and provide meaningful feedback to the user on any corrections needed.

We believe that implementing this feature will be a substantial step forward in enhancing our AI project's testing capabilities and overall efficiency. We look forward to the community's feedback and contributions to this proposal.

### PR DESCRIPTION
**Pull request by vishisht-dubey**
Hiii @Ashesh3 as per the discussion we had kindly review the pr and test whether it is working fine or not and please let me know in case of any issue. Thanks.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,4 +1,4 @@
-NEXT_PUBLIC_API_URL="https://ayushma-api.ohc.network/api/"
+NEXT_PUBLIC_API_URL="https://ayushma.ashsec.us.to/api/"
```
Comment (skks1212) [inline .env:1]: @vishisht-dubey never change `.env` files. Almost always create a new `.env.local` file to overwrite during development

File: src/app/admin/projects/[project_id]/chats/[chat_id]/page.tsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ export default function Page({
   });
   const project: Project = projectQuery.data;
   const chats: any | undefined = chatQuery.data;
-
+  console.log(project.title)
```
Comment (skks1212) [inline src/app/admin/projects/[project_id]/chats/[chat_id]/page.tsx:27]: remove unnecessary console logs

**Review by skks1212 — CHANGES_REQUESTED** [review]

**Comment by skks1212** [comment]
@vishisht-dubey also fix merge conflicts

File: .env (hunk 1 above)
Comment (vishisht-dubey) [reply .env:1]: ok sure

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey also fix merge conflicts

Hii @skks1212 i have done that can you now please review and test the code. Thanks.

File: package.json [hunk 3]
```diff
@@ -22,11 +22,13 @@
     "@types/three": "^0.160.0",
     "audio-recorder-polyfill": "^0.4.1",
     "framer-motion": "^11.0.3",
+    "fs": "^0.0.1-security",
```
Comment (Ashesh3) [inline package.json:25]: Why have you installed `fs`?

File: package.json [hunk 4]
```diff
@@ -22,11 +22,13 @@
...
     "jotai": "^2.0.4",
     "json2csv": "^6.0.0-alpha.2",
     "jspdf": "^2.5.1",
     "next": "^14.0.4",
     "npm-run-all": "^4.1.5",
+    "path": "^0.12.7",
```
Comment (Ashesh3) [inline package.json:31]: Why have you installed `path`?

File: src/app/admin/layout.tsx [hunk 5]
```diff
@@ -12,7 +12,6 @@ export default function Layout({ children }: { children: React.ReactNode }) {
   function toggleSidebar() {
     setSidebarExpanded(!isSidebarExpanded);
   }
-
```
Comment (Ashesh3) [inline src/app/admin/layout.tsx:15]: Keep the newline, this change adds no benefit.

File: src/app/admin/projects/[project_id]/chats/[chat_id]/page.tsx [hunk 6]
```diff
@@ -24,7 +24,6 @@ export default function Page({
   });
   const project: Project = projectQuery.data;
   const chats: any | undefined = chatQuery.data;
-
```
Comment (Ashesh3) [inline src/app/admin/projects/[project_id]/chats/[chat_id]/page.tsx:27]: Keep the newline, this change adds no benefit.

File: src/app/admin/tests/[testsuite_id]/page.tsx [hunk 7]
```diff
@@ -389,6 +390,51 @@ export default function Page({ params }: { params: { testsuite_id: string } }) {
     });
   };
 
+  const papaparseOptions = {
```
Comment (Ashesh3) [inline src/app/admin/tests/[testsuite_id]/page.tsx:393]: Let's call it `csvParseOptions`, it's more intuitive that way, even though I agree it's options for `papaparse` library, this variable name might not communicate what's the purpose of papaparse.

**Review by Ashesh3 — CHANGES_REQUESTED** [review]

File: src/app/admin/tests/[testsuite_id]/page.tsx (hunk 7 above)
Comment (vishisht-dubey) [reply src/app/admin/tests/[testsuite_id]/page.tsx:393]: ok

**Comment by vishisht-dubey** [comment]
Hey @Ashesh3 i have done some changes please check now

**Comment by skks1212** [comment]
@vishisht-dubey 
image
- Can you make a better UI for this?

- Can you fix this for darkmode?

- Can you share a csv file that I can upload here?

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey image
> 
> * Can you make a better UI for this?
> * Can you fix this for darkmode?
> * Can you share a csv file that I can upload here?

[tableConvert.com_0j4u1o.csv](https://github.com/coronasafe/ayushma_fe/files/14524281/tableConvert.com_0j4u1o.csv)
this is the dummy csv which i am using 

can i create a custom input element of my own for improving the UI with dark mode compatibility?

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey image
> 
> * Can you make a better UI for this?
> * Can you fix this for darkmode?
> * Can you share a csv file that I can upload here?

Screenshot 2024-03-07 184538
is this UI Ok??

**Comment by skks1212** [comment]
@vishisht-dubey if there are no premade components for uploading files, you can go ahead and create one. Although have a look at the document upload page

Also I see in the CSV languages are not in ISO language format as requested in the issue.

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey if there are no premade components for uploading files, you can go ahead and create one. Although have a look at the document upload page
> 
> Also I see in the CSV languages are not in ISO language format as requested in the issue.

can you share some csv??

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey if there are no premade components for uploading files, you can go ahead and create one. Although have a look at the document upload page
> 
> Also I see in the CSV languages are not in ISO language format as requested in the issue.



> @vishisht-dubey if there are no premade components for uploading files, you can go ahead and create one. Although have a look at the document upload page
> 
> Also I see in the CSV languages are not in ISO language format as requested in the issue.

@skks1212 is this UI ok??

**Comment by skks1212** [comment]
@Ashesh3 any comments on the UI?

**Comment by Ashesh3** [comment]
image

Let's keep a simple UI like this ^.

That button opens a file chooser where the user can then select a csv file.

**Comment by vishisht-dubey** [comment]
> image
> 
> Let's keep a simple UI like this ^.
> 
> That button opens a file chooser where the user can then select a csv file.
> 
> cc: @skks1212 what do you think about this?

so on clicking import questions a modal will open which contais that file input button right?

**Comment by skks1212** [comment]
@Ashesh3 I agree, this would look better

**Comment by Ashesh3** [comment]
> > image
> > Let's keep a simple UI like this ^.
> > That button opens a file chooser where the user can then select a csv file.
> > cc: @skks1212 what do you think about this?
> 
> so on clicking import questions a modal will open which contains that file input button right?

Use this button itself as the file upload button. No need for a modal just to show another file upload button.

https://stackoverflow.com/questions/54673206/how-to-read-and-upload-a-file-in-reactjs-using-custom-button

**Comment by vishisht-dubey** [comment]
@Ashesh3 @skks1212 I have done all the changes as suggested, can you now test the code and let me know in case of any issue. Thanks

**Comment by vishisht-dubey** [comment]
also @Ashesh3 @skks1212 
this add question modal, refer below image
Screenshot 2024-03-09 193736
see language box is not properly visible can i change its background to dark for dark mode only then it will look like this
Screenshot 2024-03-09 194118
this looks better i guess what are your opinions??

**Comment by Ashesh3** [comment]
> also @Ashesh3 @skks1212 this add question modal, refer below image Screenshot 2024-03-09 193736 see language box is not properly visible can i change its background to dark for dark mode only then it will look like this Screenshot 2024-03-09 194118 this looks better i guess what are your opinions??

Yep, bottom one looks better. Make that change.

Btw, your design looks fine except just one thing: 
Dark mode:
image

Light mode:
image

In light mode, the text is black, it should be white like other buttons. 

To change the theme:
image

**Comment by vishisht-dubey** [comment]
> > also @Ashesh3 @skks1212 this add question modal, refer below image Screenshot 2024-03-09 193736 see language box is not properly visible can i change its background to dark for dark mode only then it will look like this Screenshot 2024-03-09 194118 this looks better i guess what are your opinions??
> 
> Yep, bottom one looks better. Make that change.
Should i do that in the same pr or create an issue and then link a pr??
> 
> Btw, your design looks fine except just one thing: Dark mode: image
> 
> Light mode: image
> 
> In light mode, the text is black, it should be white like other buttons.
ya sure, i am doing that change as well
> 
> To change the theme: image

**Comment by vishisht-dubey** [comment]
> > also @Ashesh3 @skks1212 this add question modal, refer below image Screenshot 2024-03-09 193736 see language box is not properly visible can i change its background to dark for dark mode only then it will look like this Screenshot 2024-03-09 194118 this looks better i guess what are your opinions??
> 
> Yep, bottom one looks better. Make that change.
should i create a separate issue for that or do it in same pr??

**Comment by Ashesh3** [comment]
@vishisht-dubey as you feel, you can do it in the same PR or create a new issue :)

**Comment by vishisht-dubey** [comment]
> @vishisht-dubey as you feel, you can do it in the same PR or create a new issue :)

Hiii i have done changes let's keep this issue separate with that modal, for that modal i will create a separate issue
can you please test the code now and let me know in case of any issue.Thanks.

**Comment by vishisht-dubey** [comment]
@Ashesh3 any update on this?

**Review by Ashesh3 — APPROVED** [review]
Looks good to me!

**Comment by vishisht-dubey** [comment]
> Looks good to me!

are you merging this pr?

**Comment by Ashesh3** [comment]
@skks1212 will review it and then we can merge it.

**Comment by skks1212** [comment]
PR merged, thank you for contributing! 😀

**Pull request merged**